### ISSUE TITLE
Programmatically constructed CompilationUnit doesn't set the parent chain "properly"

### ISSUE THREAD

**Issue by hheg** [javaparser/javaparser#316]
I'm not entirely sure if this is a bug or not, but it feels like if the API is having a glitch or unexpected behaviour or perhaps just me not using it correctly. I wrote below tests to show, what I think, is the problem. I ran into it while writing a visitor which is supposed to "walk back up" the AST tree to find a common entrypoint and found out there's a difference between parsed CU's and created CU's.

I'd rather have that whenever you do like fex: `ASTHelper.addTypeDeclaration(createdCU, createdClassOrInterfaceDeclaration);` parent should be set.

```
/* This test shows that it the assertion will fail when checking the parent of a ClassOrInterfaceDeclaration even if it's set explicitly. Should be working imo. */
    @Test
    public void testProbableBugWithParentsNotBeingSet_SettingCompilationUnitAsClassNodeParent() throws ParseException{
        //As Reference
        CompilationUnit parsedCompilationUnit = getParsedCompilationUnit();
        assertTrue(hasUnbrokenParentsLink(parsedCompilationUnit));

        //Trying to create a CU manually
        CompilationUnit createdCU = new CompilationUnit();
        PackageDeclaration createdPackageDeclaration = new PackageDeclaration(ASTHelper.createNameExpr("test"));
        createdCU.setPackage(createdPackageDeclaration);
        assertEquals(createdPackageDeclaration.getParentNode(),createdCU);

        //public class TestClass {
        ClassOrInterfaceDeclaration createdClassOrInterfaceDeclaration = new ClassOrInterfaceDeclaration(ModifierSet.PUBLIC,
                false, "TestClass");
        ASTHelper.addTypeDeclaration(createdCU, createdClassOrInterfaceDeclaration);
        createdClassOrInterfaceDeclaration.setParentNode(createdCU); // below assertion still fails, but it works when the others are set
//      assertEquals(createdCU,createdClassOrInterfaceDeclaration);

        // public String field;
        FieldDeclaration createdFieldDeclaration = createField(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdFieldDeclaration.getParentNode());

        // public void method(){}
        MethodDeclaration createdMethodDeclaratoin = createMethod(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdMethodDeclaratoin.getParentNode());

        assertTrue(hasUnbrokenParentsLink(createdCU));
        assertEquals(parsedCompilationUnit,createdCU);      
    }
```

```
/* This test shows that when you set the parent of a ClassOrInterfaceDeclaration the assertion is a success when you add members to the class. */
    @Test
    public void testProbableBugWithParentsNotBeingSet_SettingCompilationUnitAsClassNodeParent() throws ParseException {
        //As Reference
        CompilationUnit parsedCompilationUnit = getParsedCompilationUnit();
        assertTrue(hasUnbrokenParentsLink(parsedCompilationUnit));

        //Trying to create a CU manually
        CompilationUnit createdCU = new CompilationUnit();
        PackageDeclaration createdPackageDeclaration = new PackageDeclaration(ASTHelper.createNameExpr("test"));
        createdCU.setPackage(createdPackageDeclaration);
        assertEquals(createdPackageDeclaration.getParentNode(),createdCU);

        //public class TestClass {
        ClassOrInterfaceDeclaration createdClassOrInterfaceDeclaration = new ClassOrInterfaceDeclaration(ModifierSet.PUBLIC,
                false, "TestClass");
        ASTHelper.addTypeDeclaration(createdCU, createdClassOrInterfaceDeclaration);
        createdClassOrInterfaceDeclaration.setParentNode(createdCU); // below assertion still fails, but it works when the others are set
//      assertEquals(createdCU,createdClassOrInterfaceDeclaration);

        // public String field;
        FieldDeclaration createdFieldDeclaration = ASTHelper.createFieldDeclaration(ModifierSet.PUBLIC,
                ASTHelper.createReferenceType("String", 0), "field");
        ASTHelper.addMember(createdClassOrInterfaceDeclaration, createdFieldDeclaration);
        createdFieldDeclaration.setParentNode(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdFieldDeclaration.getParentNode());

        // public void method(){}
        MethodDeclaration createdMethodDeclaratoin = new MethodDeclaration(ModifierSet.PUBLIC, ASTHelper.VOID_TYPE, "method");
        BlockStmt bstmt = new BlockStmt();
        createdMethodDeclaratoin.setBody(bstmt);
        ASTHelper.addMember(createdClassOrInterfaceDeclaration, createdMethodDeclaratoin);
        createdMethodDeclaratoin.setParentNode(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdMethodDeclaratoin.getParentNode());

        assertTrue(hasUnbrokenParentsLink(createdCU));
        assertEquals(parsedCompilationUnit,createdCU);      
    }
```

```
/* This shows that if you don't set a ClassOrInterfaceDeclaration parent it still fails later on in the reference check */
    @Test
    public void testProbableBugWithParentsNotBeingSet_NotSettingCompilationUnitAsClassNodeParent() throws ParseException{
        //As Reference
        CompilationUnit parsedCompilationUnit = getParsedCompilationUnit();
        assertTrue(hasUnbrokenParentsLink(parsedCompilationUnit));

        //Trying to create a CU manually
        CompilationUnit createdCU = new CompilationUnit();
        PackageDeclaration createdPackageDeclaration = new PackageDeclaration(ASTHelper.createNameExpr("test"));
        createdCU.setPackage(createdPackageDeclaration);
        assertEquals(createdPackageDeclaration.getParentNode(),createdCU);

        //public class TestClass {
        ClassOrInterfaceDeclaration createdClassOrInterfaceDeclaration = new ClassOrInterfaceDeclaration(ModifierSet.PUBLIC,
                false, "TestClass");
        ASTHelper.addTypeDeclaration(createdCU, createdClassOrInterfaceDeclaration);
//      createdClassOrInterfaceDeclaration.setParentNode(createdCU);
//      assertEquals(createdCU,createdClassOrInterfaceDeclaration);

        // public String field;
        FieldDeclaration createdFieldDeclaration = createField(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdFieldDeclaration.getParentNode());

        // public void method(){}
        MethodDeclaration createdMethodDeclaratoin = createMethod(createdClassOrInterfaceDeclaration);
        assertEquals(createdClassOrInterfaceDeclaration,createdMethodDeclaratoin.getParentNode());

        assertTrue(hasUnbrokenParentsLink(createdCU));
        assertEquals(parsedCompilationUnit,createdCU);      
    }
```

Helpers.

```
    private CompilationUnit getParsedCompilationUnit() throws ParseException{
        String file = 
                "package test;"+
                "public class TestClass {"+
                "   public String field;"+
                "   public void method(){}"+
                "}";
        InputStream is = new ByteArrayInputStream(file.getBytes(StandardCharsets.UTF_8));
        return JavaParser.parse(is);
    }

    private FieldDeclaration createField(ClassOrInterfaceDeclaration createdClassOrInterfaceDeclaration){
        FieldDeclaration createdFieldDeclaration = ASTHelper.createFieldDeclaration(ModifierSet.PUBLIC,
                ASTHelper.createReferenceType("String", 0), "field");
        ASTHelper.addMember(createdClassOrInterfaceDeclaration, createdFieldDeclaration);
        createdFieldDeclaration.setParentNode(createdClassOrInterfaceDeclaration);
        return createdFieldDeclaration;
    }

    private MethodDeclaration createMethod(ClassOrInterfaceDeclaration createdClassOrInterfaceDeclaration){
        MethodDeclaration createdMethodDeclaration = new MethodDeclaration(ModifierSet.PUBLIC, ASTHelper.VOID_TYPE, "method");
        BlockStmt bstmt = new BlockStmt();
        createdMethodDeclaration.setBody(bstmt);
        ASTHelper.addMember(createdClassOrInterfaceDeclaration, createdMethodDeclaration);
        createdMethodDeclaration.setParentNode(createdClassOrInterfaceDeclaration);
        return createdMethodDeclaration;
    }

    private boolean hasUnbrokenParentsLink(CompilationUnit cu){
        PackageDeclaration package1 = cu.getPackage();
        assertEquals(package1.getParentNode(),cu);

        TypeDeclaration classTypeDeclaration = cu.getTypes().get(0);
        assertEquals(cu,classTypeDeclaration.getParentNode());
        assertTrue(classTypeDeclaration instanceof ClassOrInterfaceDeclaration);
        ClassOrInterfaceDeclaration classOrInterfaceDeclaration = (ClassOrInterfaceDeclaration) classTypeDeclaration;
        assertEquals("TestClass",classOrInterfaceDeclaration.getName());

        BodyDeclaration fieldBodyDeclaration = classOrInterfaceDeclaration.getMembers().get(0);
        assertEquals(classOrInterfaceDeclaration, fieldBodyDeclaration.getParentNode());
        assertTrue(fieldBodyDeclaration instanceof FieldDeclaration);

        BodyDeclaration methodBodyDeclaration = classOrInterfaceDeclaration.getMembers().get(1);
        assertEquals(classOrInterfaceDeclaration,methodBodyDeclaration.getParentNode());
        assertTrue(methodBodyDeclaration instanceof MethodDeclaration);
        return true;
    }
```

JDK 1.8_45 and JavaParser v2.4.0.

**Comment by matozoid** [comment]
I put your code in here: https://github.com/matozoid/javaparser-issue-316

**Comment by matozoid** [comment]
One thing I found is a nasty little problem: when you call node.setXyz(List xyzs), then all xyzs will get "node" as their parent. However, if you do list=node.getXyz(), then add an xyz to the list, nothing special happens because it is just a simple Java list. Which is what happens in ASTHelper.

**Comment by matozoid** [comment]
The first two tests are green, by the way. Was that your intention?

**Comment by hheg** [comment]
```
        createdClassOrInterfaceDeclaration.setParentNode(createdCU); // below assertion still fails, but it works when the others are set
//      assertEquals(createdCU,createdClassOrInterfaceDeclaration);
```

The assert in both tests should be not be commentated really but to prove the point I commented them out.

**Comment by matozoid** [comment]
Alright, the ASTParser methods should now work on master. The problem with returned lists being mutable, making it possible to add nodes with a bad parent, is hard to solve and still there.

**Comment by matozoid** [comment]
This should be fixed when #421 is done, so no need to keep this issue open.
